src/main/scala/mesosphere/marathon/core/externalvolume/providers/DVDIProvider.scala

```diff
@@ -101,143 +105,142 @@ protected[externalvolume] case object DVDIProvider
           s"Number of instances is limited to 1 when declaring DVDI volumes in app ${app.id}", None
         )))
       else Success
   }
 
-  private val haveUniqueExternalVolumeNames = new Validator[AppDefinition] {
+  private[this] case object haveUniqueExternalVolumeNames extends Validator[AppDefinition] {
     override def apply(app: AppDefinition): Result = {
       val conflicts = volumeNameCounts(app).filter(_._2 > 1).keys
       if (conflicts.isEmpty) Success
       else Failure(conflicts.toSet[String].map { e =>
-        RuleViolation(app.id, s"Requested DVDI volume ${e} is declared more than once within app ${app.id}", None)
+        RuleViolation(app.id, s"Requested DVDI volume '$e' is declared more than once within app ${app.id}", None)
       })
     }
   }
 
-  private def volumesForApp(app: AppDefinition): Iterable[ExternalVolume] =
-    app.container.toSet[Container].flatMap(collect)
-
   // for now this matches the validation for resident tasks, but probably won't be as
   // restrictive in the future.
-  private val validUpgradeStrategy = validator[UpgradeStrategy] { strategy =>
+  private[this] lazy val validUpgradeStrategy = validator[UpgradeStrategy] { strategy =>
     strategy.minimumHealthCapacity should be <= 0.5
     strategy.maximumOverCapacity should equalTo(0.0)
   }
 
   /** @return a count of volume references-by-name within an app spec */
-  private def volumeNameCounts(app: AppDefinition): Map[String, Int] =
-    volumesForApp(app).flatMap{ v => nameOf(v.external) }.groupBy(identity).mapValues(_.size)
-
-  protected[providers] def modes(ct: Container): Set[Mode] =
-    collect(ct).map(_.mode).toSet
+  private[this] def volumeNameCounts(app: AppDefinition): Map[String, Int] =
+    app.externalVolumes.flatMap{ v => nameOf(v.external) }.groupBy(identity).mapValues(_.size)
 
-  private val validMesosVolume = validator[ExternalVolume] { v =>
+  private[this] lazy val validMesosVolume = validator[ExternalVolume] { v =>
     v.mode is equalTo(Mode.RW)
   }
 
-  private val validMesosContainer = validator[Container] { ct =>
+  private[this] lazy val validMesosContainer = validator[Container] { ct =>
     ct.volumes.each is ifDVDIVolume(validMesosVolume)
   }
 
-  private val haveOnlyOneDriver = new Validator[Container] {
-    override def apply(ct: Container): Result = {
-      val n = collect(ct).flatMap(_.external.options.get(driverOption)).toSet.size
+  private[this] case object haveOnlyOneDriver extends Validator[AppDefinition] {
+    override def apply(app: AppDefinition): Result = {
+      val uniqueProviderNames = app.externalVolumes.iterator.map(_.external.provider).toSet
+      val n = uniqueProviderNames.size
       if (n <= 1) Success
       else Failure(Set(RuleViolation(n, "one external volume driver per app", None)))
     }
   }
 
-  private val haveOnlyDriverOption = new Validator[ExternalVolume] {
+  private[this] case object haveOnlyDriverOption extends Validator[ExternalVolume] {
```

Comment:
basically using vals in "global" scope in Scala is dangerous. If you access them at the wrong time (during initialization), they can be null. Both `lazy val` and `case object` avoid this pitfall.